### PR TITLE
feat: Batch of changes and prototype code cleanup

### PR DESCRIPTION
**Pull request by pwnage101**
* Add a title field to subsdy models.
* Make the customer_uuid field an actual UUIDField.
* Overhaul the subsidy python API functions to simplify and correct.
* Remove the "requests" logic from the subisdy models.
* Remove all the prototype code related to the Subsidy Access Policy.
* Other code quality fixes.
* makemigrations.

ENT-6655

### CODE REVIEW

File: enterprise_subsidy/apps/subsidy/api.py [hunk 1]
```diff
@@ -58,16 +65,14 @@ def create_subscription_subsidy(
...
-    )
     ledger = create_ledger(
         unit=unit,
         idempotency_key=create_idempotency_key_for_subsidy(subsidy),
     )
 
     subsidy.ledger = ledger
-    if kwargs.get('starting_balance'):
+    # The following condition is crafted to respect 0 as an explicit starting balance.
+    if 'starting_balance' in kwargs and kwargs['starting_balance'] is not None:
```
Comment (johnnagro) [inline enterprise_subsidy/apps/subsidy/api.py:75]: similarly to the credit subsidy, i am going to suggest we make `starting_balance` an `arg` rather than a `kwarg`

File: enterprise_subsidy/apps/subsidy/models.py [hunk 2]
```diff
@@ -27,32 +29,34 @@ def now():
     return UTC.localize(datetime.utcnow())
 
 
-class TimeStampedModelWithUuid(TimeStampedModel):
```
Comment (johnnagro) [inline enterprise_subsidy/apps/subsidy/models.py:30]: why did you remove this? i presume we're going to want uuids and timestamps on most stuff? this makes it easier/consistent?

File: enterprise_subsidy/apps/subsidy/models.py [hunk 3]
```diff
@@ -27,32 +29,34 @@ def now():
...
+        """
         abstract = True
 
+    uuid = models.UUIDField(
+        primary_key=True,
+        default=uuid4,
+        editable=False,
+        unique=True,
+    )
+    # `title` is required for downstream revenue recognition.  The intention is to serve as a replacement to "offer
```
Comment (johnnagro) [inline enterprise_subsidy/apps/subsidy/models.py:53]: what is an example of a value here?

File: enterprise_subsidy/apps/subsidy/models.py [hunk 4]
```diff
@@ -27,32 +29,34 @@ def now():
...
-    uuid = models.UUIDField(
-        primary_key=True,
-        default=uuid4,
-        editable=False,
-        unique=True,
-    )
-
-
-class Subsidy(TimeStampedModelWithUuid):
+class Subsidy(TimeStampedModel):
```
Comment (johnnagro) [inline enterprise_subsidy/apps/subsidy/models.py:32]: We may want some class method helpers like `find_subsidies_for_customer_uuid`

File: enterprise_subsidy/apps/subsidy/models.py [hunk 5]
```diff
@@ -296,64 +323,99 @@ class AccessMethods:
     REQUEST = 'request'
 
 
-class SubsidyAccessPolicy(TimeStampedModelWithUuid):
+class SubsidyAccessPolicy(TimeStampedModel):
```
Comment (johnnagro) [inline enterprise_subsidy/apps/subsidy/models.py:326]: this code was part of the example code but it is meant for the subsidy access policy API, we can remove this.

File: enterprise_subsidy/apps/subsidy/models.py [hunk 6]
```diff
@@ -378,5 +442,8 @@ def is_entitled(self, learner_id, content_key):
 
 class LicenseRequestAccessPolicy(SubscriptionAccessPolicy):
```
Comment (johnnagro) [inline enterprise_subsidy/apps/subsidy/models.py:443]: this code was part of the example code but it is meant for the subsidy access policy API, we can remove this.

File: enterprise_subsidy/apps/subsidy/models.py [hunk 7]
```diff
@@ -296,64 +323,99 @@ class AccessMethods:
...
 class SubscriptionAccessPolicy(SubsidyAccessPolicy):
     """
+    A subsidy access policy for subscriptions (all you can eat).
+
+    .. no_pii:
     """
     subsidy = models.ForeignKey(SubscriptionSubsidy, null=True, on_delete=models.SET_NULL)
 
 
 class LearnerCreditAccessPolicy(SubsidyAccessPolicy):
```
Comment (johnnagro) [inline enterprise_subsidy/apps/subsidy/models.py:414]: this code was part of the example code but it is meant for the subsidy access policy API, we can remove this.

File: enterprise_subsidy/apps/subsidy/models.py [hunk 8]
```diff
@@ -296,64 +323,99 @@ class AccessMethods:
...
-            return self.subsidy.has_redeemed(learner_id, content_key)
-        if self.access_method == AccessMethods.REQUEST:
+            return self.subsidy.has_redeemed(learner_id, content_key)  # pylint: disable=no-member
+        elif self.access_method == AccessMethods.REQUEST:
             return self.subsidy.has_requested(learner_id, content_key)
+        else:
+            raise ValueError(f"unknown access method {self.access_method}")
 
 
 class SubscriptionAccessPolicy(SubsidyAccessPolicy):
```
Comment (johnnagro) [inline enterprise_subsidy/apps/subsidy/models.py:405]: this code was part of the example code but it is meant for the subsidy access policy API, we can remove this.

File: enterprise_subsidy/apps/subsidy/models.py [hunk 9]
```diff
@@ -189,22 +201,22 @@ def _create_redemption(self, learner_id, content_key, **kwargs):
...
         )
 
         try:
-            reference_id = self.enrollment_client.enroll(learner_id, content_key, transaction)
-            self.commit_transaction(transaction, reference_id)
+            reference_id = self.enrollment_client.enroll(learner_id, content_key, ledger_transaction)
+            self.commit_transaction(ledger_transaction, reference_id)
         except Exception as exc:
-            self.rollback_transaction(transaction)
+            self.rollback_transaction(ledger_transaction)
```
Comment (johnnagro) [inline enterprise_subsidy/apps/subsidy/models.py:214]: we should pass idempotentcy keys in?

File: enterprise_subsidy/apps/subsidy/models.py [hunk 10]
```diff
@@ -189,22 +201,22 @@ def _create_redemption(self, learner_id, content_key, **kwargs):
...
             idempotency_key,
             quantity * -1,
             transaction_metadata,
         )
 
         try:
-            reference_id = self.enrollment_client.enroll(learner_id, content_key, transaction)
-            self.commit_transaction(transaction, reference_id)
+            reference_id = self.enrollment_client.enroll(learner_id, content_key, ledger_transaction)
+            self.commit_transaction(ledger_transaction, reference_id)
```
Comment (johnnagro) [inline enterprise_subsidy/apps/subsidy/models.py:212]: we should pass idempotentcy keys in?

**Review by johnnagro — CHANGES_REQUESTED** [review]

File: enterprise_subsidy/apps/subsidy/models.py (hunk 2 above)
Comment (iloveagent57) [reply enterprise_subsidy/apps/subsidy/models.py:30]: It might be worth adding a `TimeStampedModelWithUuid` model to https://github.com/openedx/edx-django-utils

File: enterprise_subsidy/apps/subsidy/models.py (hunk 9 above)
Comment (pwnage101) [reply enterprise_subsidy/apps/subsidy/models.py:214]: ditto to my last comment.

File: enterprise_subsidy/apps/subsidy/models.py (hunk 3 above)
Comment (pwnage101) [reply enterprise_subsidy/apps/subsidy/models.py:53]: Here's a random sampling of 5 offer_names from the snowflake table `prod.core.dim_offer`:

```
OFFER_NAME
Coupon [8160]-Percentage-20.00 [1100]
Coupon [12091]-Absolute-450.00 [291]
Coupon [63226]-Percentage-100 [155] ENT Offer
Coupon [85123]-Percentage-100 [40]
Coupon [75299]-Percentage-100 [12] ENT Offer
```

Comment (pwnage101) [reply enterprise_subsidy/apps/subsidy/models.py:53]: Perusing the edx-ecommerce source code, the actual definition and setting of this field seems to live in the depths of the Oscar source code.  On the surface it looks like a rendering of internal metadata about the offer, which would make this type of field better suited as a property than a persisted database column. :shrug:

File: enterprise_subsidy/apps/subsidy/models.py [hunk 11]
```diff
@@ -189,22 +201,22 @@ def _create_redemption(self, learner_id, content_key, **kwargs):
...
         except Exception as exc:
-            self.rollback_transaction(transaction)
+            self.rollback_transaction(ledger_transaction)
             raise exc
 
-        return transaction
+        return ledger_transaction
 
-    def _create_request(self, learner_id, content_key):
+    def _create_request(self, learner_id, content_key, **kwargs):
```
Comment (iloveagent57) [inline enterprise_subsidy/apps/subsidy/models.py:219]: FWIW, I don't think any of this request stuff belongs on the subsidy level - it should be up to the subsidy access policy to query the `subsidy_request` app for any existing requests.  What we'll want in the subsidy app is a consumer of siganls about requests/approval that will create a transaction/enrollment.

File: enterprise_subsidy/apps/subsidy/api.py [hunk 12]
```diff
@@ -24,16 +28,14 @@ def create_learner_credit_subsidy(customer_uuid, unit, **kwargs):
...
```
[the remark this answers is not in the document]
Comment (pwnage101) [reply enterprise_subsidy/apps/subsidy/api.py:38]: This pair of functions needs overhaul anyway because starting_balance is passed as an object default to a subsidy above (but subsidies do not have a starting_balance field), and the functions do not cover all corner cases elegantly, according to my lookup table of outcomes:


```
subsidy_exists: get_or_create() gets an existing subsidy.
ledger_attached: the pre-existing subsidy had a non-null ledger FK. 
ledger_passed: a non-None ledger object was passed to the function.

subsidy_exists  ledger_attached  ledger_passed  outcome
0               n/a              0              [GOOD] no subsidy found, create subsidy with no ledger, create ledger and attach to subsidy
0               n/a              1              [GOOD] no subsidy found, create subsidy with passed ledger
1               0                0              [GOOD] subsidy found with null ledger, create ledger and attach to subsidy
1               0                1              [BAD] subsidy found with null ledger, but a ledger passed, so a subsidy with a null ledger returned
1               1                0              [BAD] subsidy found with attached ledger, but a new ledger is created anyway and the old one is orphaned
1               1                1              [WEIRD] subsidy found with attached ledger, but a potentially different ledger is passed and discarded
```

File: enterprise_subsidy/apps/subsidy/models.py [hunk 13]
```diff
@@ -165,7 +177,7 @@ def enrollment_client(self):
     def price_for_content(self, content_key):
         return self.catalog_client.get_content_metadata(content_key).get('price') * CENTS_PER_DOLLAR
 
-    def is_redeemable(self, learner_id, content_key, redemption_datetime=None):
+    def is_redeemable(self, content_key, redemption_datetime=None):
         return self.current_balance() >= self.price_for_content(content_key)
 
     def has_redeemed(self, learner_id, content_key, **kwargs):
```
Comment (pwnage101) [inline enterprise_subsidy/apps/subsidy/models.py:183]: @iloveagent57 I was tending to my TODO list and one of the items I have from you is to rename "has redeemed" to something like "does redemption exist".  However, we're returning a "redemption" which is really a transaction, and bubbling it up in callers.  I think a better name would be "get_redemption" which doesn't imply that a boolean is returned.  thoughts?

Comment (pwnage101) [reply enterprise_subsidy/apps/subsidy/models.py]: made the change to `get_redemption` as I described.

File: tox.ini [hunk 14]
```diff
@@ -47,8 +47,6 @@ commands =
 setenv = 
     DJANGO_SETTINGS_MODULE = enterprise_subsidy.settings.test
     PYTHONPATH = {toxinidir}
-    # Adding the option here instead of as a default in the docs Makefile because that Makefile is generated by shpinx.
-    SPHINXOPTS = -W
```
Comment (pwnage101) [inline tox.ini:51]: I spent like 20 minutes trying to squash a sphinx warning, but couldn't figure it out before giving up.  This change just makes us ignore sphinx warnings.

Comment (pwnage101) [reply tox.ini:51]: For future reference, here was the warning:

```
WARNING: autodoc: failed to import module 'urls' from module 'enterprise_subsidy'; the following exception was raised:
Traceback (most recent call last):
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/sphinx/ext/autodoc/importer.py", line 58, in import_module
    return importlib.import_module(modname)
  File "/home/tsankey/tmp/pyenv/versions/3.8.13/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/tsankey/tmp/edx/enterprise-subsidy/enterprise_subsidy/urls.py", line 20, in <module>
    from auth_backends.urls import oauth2_urlpatterns
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/auth_backends/urls.py", line 7, in <module>
    from auth_backends.views import (
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/auth_backends/views.py", line 10, in <module>
    from social_django.utils import load_strategy, load_backend
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/social_django/utils.py", line 15, in <module>
    Strategy = module_member(STRATEGY)
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/social_core/utils.py", line 58, in module_member
    module = import_module(mod)
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/social_core/utils.py", line 52, in import_module
    __import__(name)
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/auth_backends/strategies.py", line 7, in <module>
    from social_django.strategy import DjangoStrategy
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/social_django/strategy.py", line 4, in <module>
    from django.contrib.contenttypes.models import ContentType
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/django/contrib/contenttypes/models.py", line 133, in <module>
    class ContentType(models.Model):
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/django/db/models/base.py", line 108, in __new__
    app_config = apps.get_containing_app_config(module)
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/django/apps/registry.py", line 253, in get_containing_app_config
    self.check_apps_ready()
  File "/home/tsankey/tmp/edx/enterprise-subsidy/.tox/docs/lib/python3.8/site-packages/django/apps/registry.py", line 136, in check_apps_ready
    raise AppRegistryNotReady("Apps aren't loaded yet.")
django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
```

**Review by iloveagent57 — APPROVED** [review]
